design/speaker.md

```diff
@@ -0,0 +1,128 @@
+                                                            │  <<use>>
+                                                    ┌───────▼───────┐
+                                                    │               │
+                                                    │    session    │
+                                                    │ <<interface>> │
+                                                    │               │
+                                                    └───────▲───────┘
+                                                            │ <<implements>>
+                                                            │
+                                                    ┌───────┴───────┐
+                                                    │               │
+                                                    │     bgp       │
+                                                    │  <<package>>  │
+                                                    │               │
+                                                    └───────────────┘
+
+```
+
+### main
+
+The main package contains the main() entrypoint for the speaker application. It
+creates an L2 and/or BGP controller, a speakerList and a k8s client. The k8s
+client watches for changes to the MetalLB config map, service configuration, and
+the node configuration and calls the appropriate protocol handler (e.g. L2, BGP)
+via the “Protocol” Interface after ensuring the change is valid.
+
+The BGP implementation of the “Protocol” interface can be found
+[here](https://github.com/metallb/metallb/blob/main/speaker/bgp_controller.go).
+This handles changes to the config map, service, and node configuration for BGP
+pools. Based on these changes, BGP sessions will be established/destroyed and
+prefixes will be advertised via the bgp package.
+
+The Layer 2 implementation of the "Protocol" interface can be found
+[here](https://github.com/metallb/metallb/blob/main/speaker/layer2_controller.go).
+
+### bgp
+
+The bgp package manages BGP sessions and advertisements. A BGP session is
+created by the New() method and prefixes are advertised by the Set() method
+which is part of the session interface.
+
+BGP sessions are uniquely identified and configured by the parameters of the
+New() method which are currently:
+
+* **addr**: IP address and port of peer
+* **asn**: Local ASN
+* **routerID**: Specify the BGP Router ID of the session
+* **peerASN**: ASN of the router that we are connecting to
+* **holdTime**: BGP Hold Time specifies how long to maintain sessions for
+* **Password**: TCP MD5 sock option for connection
+* **myNode**
+
+Advertisements are made for a session via the Set() method.
+
+### speakerlist
+
+speakerlist uses the Hashicorp memberlist package to create a cluster of
+speakers, manage membership of this cluster and to detect member failures using
+the “gossip” protocol. This is only used in the speaker's L2 mode.
```

Comment:
```suggestion
a gossip protocol. This is only used in the speaker's L2 mode.
```

"Gossip protocols" is a category of protocols rather than the name of a specific protocol.